Update useEffectEvent to be a better match for the native implementation

Ayc0 opened the pull request:
👋 

I tried out the hook `useEffectEvent` (the actual experimental one from 'react'), and I noticed 2 differences:

1. it is not referentially stable, meaning that if passed to a `useCallback` or a `useEffect`, it'll cause those to be re-computed. `useEffectEvent` creates a new function at every render (without a wrapping `useCallback`), but this ponyfill does

   https://github.com/user-attachments/assets/a05f533d-2b48-493a-b9f8-9cdc5af9d7a8

2. when called during **any** renders (not just the 1st one), it crashes
   
   https://github.com/user-attachments/assets/d139b72d-79c8-4017-af9e-a09752dcc776


To note: I did **not** test the changes I'm proposing here 😅. I just noticed those behavioral differences between the 2 hooks and wanted to give a headsup

Ayc0 commented on src/useEffectEvent.ts:
```diff
@@ -1,5 +1,9 @@
 import {useCallback, useInsertionEffect, useRef} from 'react'
 
+function forbiddenInRender() {
+  throw new Error("A function wrapped in useEffectEvent cannot be called during rendering")
```
```suggestion
  throw new Error("A function wrapped in useEffectEvent can't be called during rendering.")
```

Ayc0 commented on src/useEffectEvent.ts:
```diff
@@ -1,5 +1,9 @@
 import {useCallback, useInsertionEffect, useRef} from 'react'
 
+function forbiddenInRender() {
+  throw new Error("A function wrapped in useEffectEvent can't be called during rendering.")
+}
```
This is to replicate the behavior [on mount](https://github.com/facebook/react/blob/cd90a4d8c0d5dbaa8ab61e839b112b1518d5058f/packages/react-reconciler/src/ReactFiberHooks.js#L2867-L2871), and [on update](https://github.com/facebook/react/blob/cd90a4d8c0d5dbaa8ab61e839b112b1518d5058f/packages/react-reconciler/src/ReactFiberHooks.js#L2884-L2888)

Ayc0 commented on src/useEffectEvent.ts:
```diff
@@ -13,12 +17,14 @@ export function useEffectEvent<
   ) => void,
 >(fn: T): T {
   const ref = useRef<T | null>(null)
+  ref.current = forbiddenInRender
```
This is technically not concurrent safe (we shouldn't be writing in refs during the render cycle), but here it's a bit the goal

`useEffectEvent` is defined in 3 different places:
- On [mount](https://github.com/facebook/react/blob/cd90a4d8c0d5dbaa8ab61e839b112b1518d5058f/packages/react-reconciler/src/ReactFiberHooks.js#L3971-L3973)
- On [update](https://github.com/facebook/react/blob/cd90a4d8c0d5dbaa8ab61e839b112b1518d5058f/packages/react-reconciler/src/ReactFiberHooks.js#L4004-L4006)
- On [rerender](https://github.com/facebook/react/blob/cd90a4d8c0d5dbaa8ab61e839b112b1518d5058f/packages/react-reconciler/src/ReactFiberHooks.js#L4038-L4040)

Both `update` & `rerender` assign it to [`updateEvent`](https://github.com/facebook/react/blob/cd90a4d8c0d5dbaa8ab61e839b112b1518d5058f/packages/react-reconciler/src/ReactFiberHooks.js#L2876-L2891) which itself checks for `isInvalidExecutionContextForEventFunction()`, doing a check on the [`RenderContext`](https://github.com/facebook/react/blob/cd90a4d8c0d5dbaa8ab61e839b112b1518d5058f/packages/react-reconciler/src/ReactFiberWorkLoop.js#L1749-L1752)

stipsan reviewed (commented):
Thanks for the PR! 💖 
I need to run some updates on this repo and setup a testing suite before we can handle this change. 
I agree we should match the native implementation here. The reason we chose to run with a stable reference for the callback were due to repos consuming `use-effect-event` being unable to easily switch to `eslint-plugin-react-hooks@experimental` and thus protect against accidentally having it in deps, or passing it as a prop.
This is no longer the case, and we've moved to `eslint-plugin-react-hooks@experimental`, and we've seen the need to protect against the "called during rendering" pitfall, so I was delighted to see you've already sent a PR for us 😌 ✨ 

I'll setup a test suite that we can rebase this PR onto, to ensure the ponyfill exactly matches the native implementation 🙌

Ayc0 commented:
> I'll setup a test suite that we can rebase this PR onto, to ensure the ponyfill exactly matches the native implementation 🙌

@stipsan, if you want, I can work on this test suite. If you have any specific guidelines on what you'd want to see in those tests, feel free to mention them 👍

stipsan commented:
> > I'll setup a test suite that we can rebase this PR onto, to ensure the ponyfill exactly matches the native implementation 🙌
> 
> 
> 
> @stipsan, if you want, I can work on this test suite. If you have any specific guidelines on what you'd want to see in those tests, feel free to mention them 👍 

If you have the time I'd be greatly appreciated 🙏❤️

Ayc0 commented:
@stipsan I did some changes to the code:

- I installed `vitest` & `@testing-library/react` & `jsdom` to be able to run tests
- I added a few tests
- I updated `eslint-plugin-react-hooks` to the experimental branch, so that `useEventEffect` could be properly omitted from dependency arrays

Ayc0 commented on .eslintrc.json:
```diff
@@ -5,6 +5,7 @@
   },
   "parser": "@typescript-eslint/parser",
   "extends": ["plugin:@typescript-eslint/recommended", "prettier"],
+  "reportUnusedDisableDirectives": true,
```
This is to avoid having unused `eslint-disable`

I can remove it from this PR if you want

stipsan reviewed (approved):
Thank you so much! Looks ready to go 💖
